### PR TITLE
[ISSUE-476] refactor read apis to support read from hdfs 

### LINKED ISSUE #476: refactor read apis to support read from hdfs.

### PR DESCRIPTION
**Pull request by FMX**
# [ISSUE-476] refactor read apis to support read from hdfs 

### Related issues.
closes #476 

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: client/src/main/java/com/aliyun/emr/rss/client/ShuffleClient.java [hunk 1]
```diff
@@ -79,6 +83,22 @@ public static ShuffleClient get(String driverHost, int port, RssConf conf) {
     return _instance;
   }
 
+  public static FileSystem getHdfsFs(RssConf conf) {
+    if (null == hdfsFs) {
+      synchronized (ShuffleClient.class) {
+        Configuration hdfsConfiguration = new Configuration();
+        hdfsConfiguration.set("fs.defaultFS", RssConf.hdfsDir(conf));
```
Comment (AngersZhuuuu) [inline client/src/main/java/com/aliyun/emr/rss/client/ShuffleClient.java:90]: Should we get this Configuration from Spark?

**Review by AngersZhuuuu — COMMENTED** [review]
LGTM

File: client/src/main/java/com/aliyun/emr/rss/client/read/RemotePartitionReader.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        exception.set(new IOException(errorMsg, e));
+      }
+    };
+    client = new RetryingChunkClient(conf, shuffleKey, location,
+      callback, clientFactory, startMapIndex, endMapIndex);
+    numChunks = client.openChunks();
+  }
+
+  public boolean hasData() {
+    return returnedChunks < numChunks && numChunks >= 1;
```
Comment (waitinfuture) [inline client/src/main/java/com/aliyun/emr/rss/client/read/RemotePartitionReader.java:86]: No need for numChunks >=1. returnedChunks default to 0, If num Chunks is 0, returnedChunks < numChunks will be false

File: client/src/main/java/com/aliyun/emr/rss/client/read/RemotePartitionReader.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        logger.error(errorMsg, e);
+        exception.set(new IOException(errorMsg, e));
+      }
+    };
+    client = new RetryingChunkClient(conf, shuffleKey, location,
+      callback, clientFactory, startMapIndex, endMapIndex);
+    numChunks = client.openChunks();
+  }
+
+  public boolean hasData() {
```
Comment (waitinfuture) [inline client/src/main/java/com/aliyun/emr/rss/client/read/RemotePartitionReader.java:85]: I think hasNext is good

File: client/src/main/java/com/aliyun/emr/rss/client/read/RssInputStream.java [hunk 4]
```diff
@@ -158,19 +149,20 @@ private void moveToNextReader() throws IOException {
...
-      logger.debug("Moved to next partition {},startMapIndex {} endMapIndex {} , {}/{} read , " +
-                    "get chunks size {}", locations[fileIndex], startMapIndex, endMapIndex,
-        fileIndex, locations.length, currentReader.numChunks);
-      while (currentReader.numChunks < 1 && fileIndex < locations.length - 1) {
+      int locationCount = locations.length;
+      PartitionLocation currentLocation = locations[fileIndex];
+      currentReader = createReader(currentLocation);
+      logger.debug("Moved to next partition {},startMapIndex {} endMapIndex {} , {}/{} read ",
+          currentLocation, startMapIndex, endMapIndex, fileIndex, locationCount);
+      while (currentReader.hasData() && fileIndex < locationCount - 1) {
```
Comment (waitinfuture) [inline client/src/main/java/com/aliyun/emr/rss/client/read/RssInputStream.java:157]: Should be !hasData()

File: client/src/main/java/com/aliyun/emr/rss/client/read/RemotePartitionReader.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+import org.slf4j.LoggerFactory;
+
+import com.aliyun.emr.rss.common.RssConf;
+import com.aliyun.emr.rss.common.network.buffer.ManagedBuffer;
+import com.aliyun.emr.rss.common.network.buffer.NettyManagedBuffer;
+import com.aliyun.emr.rss.common.network.client.ChunkReceivedCallback;
+import com.aliyun.emr.rss.common.network.client.TransportClientFactory;
+import com.aliyun.emr.rss.common.protocol.PartitionLocation;
+
+public class RemotePartitionReader implements PartitionReader {
```
Comment (waitinfuture) [inline client/src/main/java/com/aliyun/emr/rss/client/read/RemotePartitionReader.java:36]: WorkerPartitionReader is better

File: client/src/main/java/com/aliyun/emr/rss/client/ShuffleClient.java [hunk 6]
```diff
@@ -79,6 +83,21 @@ public static ShuffleClient get(String driverHost, int port, RssConf conf) {
     return _instance;
   }
 
+  public static FileSystem getHdfsFs(RssConf conf) {
+    if (null == hdfsFs) {
+      synchronized (ShuffleClient.class) {
+        Configuration hdfsConfiguration = new Configuration();
```
Comment (waitinfuture) [inline client/src/main/java/com/aliyun/emr/rss/client/ShuffleClient.java:89]: Should check null before this line again

File: common/src/main/java/com/aliyun/emr/rss/common/protocol/StorageInfo.java [hunk 7]
```diff
@@ -73,25 +82,38 @@ public void setMountPoint(String mountPoint) {
...
-  public static PbStorageHint toPb(StorageInfo hint) {
-    return PbStorageHint.newBuilder()
-             .setType(hint.type.type)
-             .setFinalResult(hint.finalResult)
-             .setMountPoint(hint.mountPoint)
-             .build();
+  public static PbStorageInfo toPb(StorageInfo storageInfo) {
+    String filePath = storageInfo.getFilePath();
+    PbStorageInfo.Builder builder = PbStorageInfo.newBuilder();
+    builder.setType(storageInfo.type.type)
```
Comment (waitinfuture) [inline common/src/main/java/com/aliyun/emr/rss/common/protocol/StorageInfo.java:106]: Better to rename the second type to value

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**